The job **`i686-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/5705510679/job/15460257340) [(plain)](https://github.com/rust-lang-ci/rust/commit/876738e067ec747bd6693499643c334d27069ca4/checks/15460257340/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] coretests test:true 112.843
[RUSTC-TIMING] alloc test:true 115.382
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x86\\link.exe" "/NOLOGO" "/LARGEADDRESSAWARE" "/SAFESEH" "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\rustc5QCApo\\symbols.o" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\std-1bf831bc50b8ddbd.std.d7a36f461044e6b9-cgu.0.rcgu.o" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\std-1bf831bc50b8ddbd.1b0yzr5v81wgt7mp.rcgu.o" "/LIBPATH:C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps" "/LIBPATH:C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\release\\deps" "/LIBPATH:C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\atlmfc\\lib\\x86" "/LIBPATH:C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\build\\compiler_builtins-777a9e98a067c7da\\out" "/LIBPATH:C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "kernel32.lib" "advapi32.lib" "bcrypt.lib" "kernel32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "kernel32.lib" "ws2_32.lib" "kernel32.lib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\librand_xorshift-41757b6698f02199.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\librand-3a57ffd4fbcc3748.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\librand_core-d33bbc29685528b4.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libtest-1e17edef25bd1f26.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libgetopts-43b695067bd205c4.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libunicode_width-c1a2973c23c42499.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\librustc_std_workspace_std-c6823333a8ba0922.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libstd-d49ee14d9eae7f42.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libpanic_unwind-2220f5946183b9c6.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\librustc_demangle-e330235f39b4c77a.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libstd_detect-3d6c68d6ce87d257.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libhashbrown-6f20bbe1a1a1b4cd.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\librustc_std_workspace_alloc-a128c25ef54ae2a8.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libminiz_oxide-a59894af7a4a9407.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libadler-408f5c4b935e000d.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libunwind-3a83e5b4a1db0fa2.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libcfg_if-ba35f73185b3cbce.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\liblibc-3eb0541ae38cc82f.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\liballoc-7c62780a30d4039b.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\librustc_std_workspace_core-795d1c39a53f4bd9.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\libcore-9cd71118615e40fd.rlib" "C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\libcompiler_builtins-db7c8ccd13d5978f.rlib" "kernel32.lib" "kernel32.lib" "advapi32.lib" "bcrypt.lib" "kernel32.lib" "ntdll.lib" "userenv.lib" "ws2_32.lib" "kernel32.lib" "ws2_32.lib" "kernel32.lib" "libcmt.lib" "/NXCOMPAT" "/LIBPATH:C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "/OUT:C:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage1-std\\i686-pc-windows-msvc\\release\\deps\\std-1bf831bc50b8ddbd.exe" "/OPT:REF,ICF" "/DEBUG"
  = note: std-1bf831bc50b8ddbd.std.d7a36f461044e6b9-cgu.0.rcgu.o : error LNK2019: unresolved external symbol _lgamma_r referenced in function __ZN4core3ops8function6FnOnce9call_once17h8b792f0a7b73def5E

          std-1bf831bc50b8ddbd.std.d7a36f461044e6b9-cgu.0.rcgu.o : error LNK2019: unresolved external symbol _lgammaf_r referenced in function __ZN4core3ops8function6FnOnce9call_once17hd28387cd023412ebE

          C:\a\rust\rust\build\i686-pc-windows-msvc\stage1-std\i686-pc-windows-msvc\release\deps\std-1bf831bc50b8ddbd.exe : fatal error LNK1120: 2 unresolved externals
          

[RUSTC-TIMING] std test:true 167.251
error: could not compile `std` (lib test) due to previous error
error: could not compile `std` (lib test) due to previous error
Build completed unsuccessfully in 0:47:10
make: *** [Makefile:66: ci-msvc-ps1] Error 1
  local time: Sun Jul 30 10:41:52 CUT 2023
  network time: Sun, 30 Jul 2023 10:41:52 GMT
##[endgroup]
##[error]Process completed with exit code 2.
```

</details>

--- a/library/std/src/f32.rs
+++ b/library/std/src/f32.rs
@@ -957,4 +957,46 @@ impl f32 {
     pub fn atanh(self) -> f32 {
         0.5 * ((2.0 * self) / (1.0 - self)).ln_1p()
     }
+
+    /// Gamma function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(float_gamma)]
+    /// let x = 5.0f32;
+    ///
+    /// let abs_difference = (x.gamma() - 24.0).abs();
+    ///
+    /// assert!(abs_difference <= f32::EPSILON);
+    /// ```
+    #[rustc_allow_incoherent_impl]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[unstable(feature = "float_gamma", issue = "99842")]
+    #[inline]
+    pub fn gamma(self) -> f32 {
+        unsafe { cmath::tgammaf(self) }
+    }
+
+    /// Returns the natural logarithm of the gamma function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(float_gamma)]
+    /// let x = 2.0f32;
+    ///
+    /// let abs_difference = (x.ln_gamma().0 - 0.0).abs();
+    ///
+    /// assert!(abs_difference <= f32::EPSILON);
+    /// ```
+    #[rustc_allow_incoherent_impl]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[unstable(feature = "float_gamma", issue = "99842")]
+    #[inline]
+    pub fn ln_gamma(self) -> (f32, i32) {
+        let mut signgamp: i32 = 0;
+        let x = unsafe { cmath::lgammaf_r(self, &mut signgamp) };
+        (x, signgamp)
+    }
 }

--- a/library/std/src/f32/tests.rs
+++ b/library/std/src/f32/tests.rs
@@ -653,6 +653,38 @@ fn test_atanh() {
 }
 
 #[test]
+fn test_gamma() {
+    // precision can differ between platforms
+    assert_approx_eq!(1.0f32.gamma(), 1.0f32);
+    assert_approx_eq!(2.0f32.gamma(), 1.0f32);
+    assert_approx_eq!(3.0f32.gamma(), 2.0f32);
+    assert_approx_eq!(4.0f32.gamma(), 6.0f32);
+    assert_approx_eq!(5.0f32.gamma(), 24.0f32);
+    assert_approx_eq!(0.5f32.gamma(), consts::PI.sqrt());
+    assert_approx_eq!((-0.5f32).gamma(), -2.0 * consts::PI.sqrt());
+    assert_eq!(0.0f32.gamma(), f32::INFINITY);
+    assert_eq!((-0.0f32).gamma(), f32::NEG_INFINITY);
+    assert!((-1.0f32).gamma().is_nan());
+    assert!((-2.0f32).gamma().is_nan());
+    assert!(f32::NAN.gamma().is_nan());
+    assert!(f32::NEG_INFINITY.gamma().is_nan());
+    assert_eq!(f32::INFINITY.gamma(), f32::INFINITY);
+    assert_eq!(171.71f32.gamma(), f32::INFINITY);
+}
+
+#[test]
+fn test_ln_gamma() {
+    assert_approx_eq!(1.0f32.ln_gamma().0, 0.0f32);
+    assert_eq!(1.0f32.ln_gamma().1, 1);
+    assert_approx_eq!(2.0f32.ln_gamma().0, 0.0f32);
+    assert_eq!(2.0f32.ln_gamma().1, 1);
+    assert_approx_eq!(3.0f32.ln_gamma().0, 2.0f32.ln());
+    assert_eq!(3.0f32.ln_gamma().1, 1);
+    assert_approx_eq!((-0.5f32).ln_gamma().0, (2.0 * consts::PI.sqrt()).ln());
+    assert_eq!((-0.5f32).ln_gamma().1, -1);
+}
+
+#[test]
 fn test_real_consts() {
     use super::consts;
 

--- a/library/std/src/f64.rs
+++ b/library/std/src/f64.rs
@@ -957,4 +957,46 @@ impl f64 {
     pub fn atanh(self) -> f64 {
         0.5 * ((2.0 * self) / (1.0 - self)).ln_1p()
     }
+
+    /// Gamma function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(float_gamma)]
+    /// let x = 5.0f64;
+    ///
+    /// let abs_difference = (x.gamma() - 24.0).abs();
+    ///
+    /// assert!(abs_difference <= f64::EPSILON);
+    /// ```
+    #[rustc_allow_incoherent_impl]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[unstable(feature = "float_gamma", issue = "99842")]
+    #[inline]
+    pub fn gamma(self) -> f64 {
+        unsafe { cmath::tgamma(self) }
+    }
+
+    /// Returns the natural logarithm of the gamma function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(float_gamma)]
+    /// let x = 2.0f64;
+    ///
+    /// let abs_difference = (x.ln_gamma().0 - 0.0).abs();
+    ///
+    /// assert!(abs_difference <= f64::EPSILON);
+    /// ```
+    #[rustc_allow_incoherent_impl]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[unstable(feature = "float_gamma", issue = "99842")]
+    #[inline]
+    pub fn ln_gamma(self) -> (f64, i32) {
+        let mut signgamp: i32 = 0;
+        let x = unsafe { cmath::lgamma_r(self, &mut signgamp) };
+        (x, signgamp)
+    }
 }

--- a/library/std/src/f64/tests.rs
+++ b/library/std/src/f64/tests.rs
@@ -636,6 +636,38 @@ fn test_atanh() {
 }
 
 #[test]
+fn test_gamma() {
+    // precision can differ between platforms
+    assert_approx_eq!(1.0f64.gamma(), 1.0f64);
+    assert_approx_eq!(2.0f64.gamma(), 1.0f64);
+    assert_approx_eq!(3.0f64.gamma(), 2.0f64);
+    assert_approx_eq!(4.0f64.gamma(), 6.0f64);
+    assert_approx_eq!(5.0f64.gamma(), 24.0f64);
+    assert_approx_eq!(0.5f64.gamma(), consts::PI.sqrt());
+    assert_approx_eq!((-0.5f64).gamma(), -2.0 * consts::PI.sqrt());
+    assert_eq!(0.0f64.gamma(), f64::INFINITY);
+    assert_eq!((-0.0f64).gamma(), f64::NEG_INFINITY);
+    assert!((-1.0f64).gamma().is_nan());
+    assert!((-2.0f64).gamma().is_nan());
+    assert!(f64::NAN.gamma().is_nan());
+    assert!(f64::NEG_INFINITY.gamma().is_nan());
+    assert_eq!(f64::INFINITY.gamma(), f64::INFINITY);
+    assert_eq!(171.71f64.gamma(), f64::INFINITY);
+}
+
+#[test]
+fn test_ln_gamma() {
+    assert_approx_eq!(1.0f64.ln_gamma().0, 0.0f64);
+    assert_eq!(1.0f64.ln_gamma().1, 1);
+    assert_approx_eq!(2.0f64.ln_gamma().0, 0.0f64);
+    assert_eq!(2.0f64.ln_gamma().1, 1);
+    assert_approx_eq!(3.0f64.ln_gamma().0, 2.0f64.ln());
+    assert_eq!(3.0f64.ln_gamma().1, 1);
+    assert_approx_eq!((-0.5f64).ln_gamma().0, (2.0 * consts::PI.sqrt()).ln());
+    assert_eq!((-0.5f64).ln_gamma().1, -1);
+}
+
+#[test]
 fn test_real_consts() {
     use super::consts;
     let pi: f64 = consts::PI;

--- a/library/std/src/lib.rs
+++ b/library/std/src/lib.rs
@@ -286,6 +286,7 @@
 #![feature(exact_size_is_empty)]
 #![feature(exclusive_wrapper)]
 #![feature(extend_one)]
+#![feature(float_gamma)]
 #![feature(float_minimum_maximum)]
 #![feature(float_next_up_down)]
 #![feature(hasher_prefixfree_extras)]

--- a/library/std/src/sys/unix/cmath.rs
+++ b/library/std/src/sys/unix/cmath.rs
@@ -30,4 +30,8 @@ extern "C" {
     pub fn tanf(n: f32) -> f32;
     pub fn tanh(n: f64) -> f64;
     pub fn tanhf(n: f32) -> f32;
+    pub fn tgamma(n: f64) -> f64;
+    pub fn tgammaf(n: f32) -> f32;
+    pub fn lgamma_r(n: f64, s: &mut i32) -> f64;
+    pub fn lgammaf_r(n: f32, s: &mut i32) -> f32;
 }

--- a/library/std/src/sys/windows/cmath.rs
+++ b/library/std/src/sys/windows/cmath.rs
@@ -1,6 +1,6 @@
 #![cfg(not(test))]
 
-use libc::{c_double, c_float};
+use libc::{c_double, c_float, c_int};
 
 extern "C" {
     pub fn acos(n: c_double) -> c_double;
@@ -23,6 +23,10 @@ extern "C" {
     pub fn sinh(n: c_double) -> c_double;
     pub fn tan(n: c_double) -> c_double;
     pub fn tanh(n: c_double) -> c_double;
+    pub fn tgamma(n: c_double) -> c_double;
+    pub fn tgammaf(n: c_float) -> c_float;
+    pub fn lgamma_r(n: c_double, s: &mut c_int) -> c_double;
+    pub fn lgammaf_r(n: c_float, s: &mut c_int) -> c_float;
 }
 
 pub use self::shims::*;

